pkg/addonmanager/manager.go

```diff
@@ -189,144 +137,9 @@ func (a *addonManager) Start(ctx context.Context) error {
-	// This should be removed when the migration is done.
-	// The migration plan refer to https://github.com/open-cluster-management-io/ocm/issues/355.
-	managementAddonController := cmamanagedby.NewCMAManagedByController(
-		addonClient,
-		addonInformers.Addon().V1alpha1().ClusterManagementAddOns(),
-		a.addonAgents,
-		utils.FilterByAddonName(a.addonAgents),
-	)
-
-	var addonConfigController, managementAddonConfigController factory.Controller
-	if len(a.addonConfigs) != 0 {
-		addonConfigController = addonconfig.NewAddonConfigController(
-			addonClient,
-			addonInformers.Addon().V1alpha1().ManagedClusterAddOns(),
-			addonInformers.Addon().V1alpha1().ClusterManagementAddOns(),
-			dynamicInformers,
-			a.addonConfigs,
-			utils.FilterByAddonName(a.addonAgents),
-		)
-		managementAddonConfigController = cmaconfig.NewCMAConfigController(
-			addonClient,
-			addonInformers.Addon().V1alpha1().ClusterManagementAddOns(),
-			dynamicInformers,
-			a.addonConfigs,
-			utils.FilterByAddonName(a.addonAgents),
-		)
-	}
-
-	var csrApproveController factory.Controller
-	var csrSignController factory.Controller
-	// Spawn the following controllers only if v1 CSR api is supported in the
-	// hub cluster. Under v1beta1 CSR api, all the CSR objects will be signed
-	// by the kube-controller-manager so custom CSR controller should be
-	// disabled to avoid conflict.
-	if v1CSRSupported {
-		csrApproveController = certificate.NewCSRApprovingController(
-			kubeClient,
-			clusterInformers.Cluster().V1().ManagedClusters(),
-			kubeInformers.Certificates().V1().CertificateSigningRequests(),
-			nil,
-			addonInformers.Addon().V1alpha1().ManagedClusterAddOns(),
-			a.addonAgents,
-		)
-		csrSignController = certificate.NewCSRSignController(
-			kubeClient,
-			clusterInformers.Cluster().V1().ManagedClusters(),
-			kubeInformers.Certificates().V1().CertificateSigningRequests(),
-			addonInformers.Addon().V1alpha1().ManagedClusterAddOns(),
-			a.addonAgents,
-		)
-	} else if v1beta1Supported {
-		csrApproveController = certificate.NewCSRApprovingController(
-			kubeClient,
-			clusterInformers.Cluster().V1().ManagedClusters(),
-			nil,
-			kubeInformers.Certificates().V1beta1().CertificateSigningRequests(),
-			addonInformers.Addon().V1alpha1().ManagedClusterAddOns(),
-			a.addonAgents,
-		)
-	}
-
-	a.syncContexts = append(a.syncContexts, deployController.SyncContext())
-
-	go deployController.Run(ctx, 1)
-	go registrationController.Run(ctx, 1)
-	go managementAddonController.Run(ctx, 1)
-
-	if addonConfigController != nil {
-		go addonConfigController.Run(ctx, 1)
-	}
-	if managementAddonConfigController != nil {
-		go managementAddonConfigController.Run(ctx, 1)
-	}
-	if csrApproveController != nil {
-		go csrApproveController.Run(ctx, 1)
-	}
-	if csrSignController != nil {
-		go csrSignController.Run(ctx, 1)
-	}
-	return nil
-}
-
 // New returns a new Manager for creating addon agents.
 func New(config *rest.Config) (AddonManager, error) {
```

Comment:
Not sure if it is better to provider a addon manager factory func, like:
```
func New(mtype AddonManagerType)(AddonManager, error){
  if mtype =="cloudevents"{
    return ...
  }
  ...
}
```

Reply:
this will break existing code that using `addonmanager.New(Config)` to create addon manager, right?
we don't want make any change to current code, that's why we keep this unchanged.